tutoraspects/templates/aspects/apps/aspects/migrations/alembic/versions/0028_user_pii.py

```diff
@@ -0,0 +1,134 @@
+"""
+create user_pii table sourced from an in-memory dictionary joining the PII tables.
+"""
+from alembic import op
+
+
+revision = "0028"
+down_revision = "0027"
+branch_labels = None
+depends_on = None
+on_cluster = " ON CLUSTER '{{CLICKHOUSE_CLUSTER_NAME}}' " if "{{CLICKHOUSE_CLUSTER_NAME}}" else ""
+engine = "ReplicatedReplacingMergeTree" if "{{CLICKHOUSE_CLUSTER_NAME}}" else "ReplacingMergeTree"
+
+def upgrade():
+
+    # We include these drop statements here because "CREATE OR REPLACE DICTIONARY"
+    # currently throws a file rename error and you can't drop a dictionary with a
+    # table referring to it.
+    op.execute(
+        f"""
+        DROP TABLE IF EXISTS {{ ASPECTS_EVENT_SINK_DATABASE }}.user_pii
+        {on_cluster}
+        """
+    )
+    op.execute(
+        f"""
+        DROP DICTIONARY IF EXISTS {{ ASPECTS_EVENT_SINK_DATABASE }}.user_pii_dict
+        {on_cluster}
+        """
+    )
+    op.execute(
+        f"""
+        CREATE DICTIONARY {{ ASPECTS_EVENT_SINK_DATABASE }}.user_pii_dict
+        {on_cluster}
+        (
+            external_user_id UUID,
+            external_id_type String,
+            username String,
+            user_id Int32,
+            name String,
+            meta String,
+            courseware String,
+            language String,
+            location String,
+            year_of_birth String,
+            gender String,
+            level_of_education String,
+            mailing_address String,
+            city String,
+            country String,
+            state String,
+            goals String,
+            bio String,
+            profile_image_uploaded_at String,
+            phone_number String
+        )
+        PRIMARY KEY user_id
+        SOURCE(CLICKHOUSE(
+            user '{{ CLICKHOUSE_ADMIN_USER }}'
+            password '{{ CLICKHOUSE_ADMIN_PASSWORD }}'
+            db '{{ ASPECTS_EVENT_SINK_DATABASE }}'
+            query "
+                with most_recent_user_profile as (
```

Comment:
I think this will fan out, I don't see any limiting that only pulls the most recent profile or most recent external id in either of these queries?

Reply:
Hmm, I didn't know external_id's could be changed. I'm reasonably sure they are currently immutable (or at least nothing updates them) 1:1 with user id / type. If nothing else we want to limit the `external_id_type` to `xapi`, but if there's a place where secret key is involved can you point it out? That would complicate several things in unexpected ways. :)

Reply:
> Hmm, I didn't know external_id's could be changed. I'm reasonably sure they are currently immutable (or at least nothing updates them) 1:1 with user id / type.

Ah, my apologies.. I was thinking of the [old anonymous user ids](https://github.com/openedx/edx-platform/blob/master/common/djangoapps/student/docs/decisions/0001-anonymous-user-id-generation.rst#L17), which were hashed on the Django app `SECRET_KEY`. External IDs are immutable, as you expected.

> If nothing else we want to limit the external_id_type to xapi

Ok that makes sense. I wasn't sure if there were any other types of external_ids used in our events. Will update the query that generates the dict/table.

Reply:
Please see [57a4f56](https://github.com/openedx/tutor-contrib-aspects/pull/529/commits/57a4f563abf38c446b01417a2e3fc10b36aa63a5).

Reply:
FYI I also added a dataset for the new user_pii table with [9ceb57b](https://github.com/openedx/tutor-contrib-aspects/pull/529/commits/9ceb57b7840a6712786768097be146d224e602d1).